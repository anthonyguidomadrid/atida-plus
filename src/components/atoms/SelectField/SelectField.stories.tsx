import { SelectField, SelectFieldProps } from '.'

type OptionValue = string
const OPTIONS = [
  { label: 'Option 1', value: 'Option 1' },
  { label: 'Option 2', value: 'Option 2' }
]

export default {
  component: SelectField,
  title: 'atoms/SelectField',
  args: {
    options: OPTIONS
  }
}

export const Basic = (args: SelectFieldProps<OptionValue>): JSX.Element => {
  return <SelectField {...args} value={OPTIONS[0].value} />
}
