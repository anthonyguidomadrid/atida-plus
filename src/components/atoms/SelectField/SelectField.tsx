import React from 'react'
import Select, { components } from 'react-select'
import type {
  ActionMeta,
  PropsValue,
  MultiValue,
  DropdownIndicatorProps,
  SingleValue,
  OnChangeValue
} from 'react-select'
import { ReactComponent as MenuDown } from '~assets/svg/MenuDown.svg'
import classNames from 'classnames'
import type { FormikProps, FieldInputProps } from 'formik'

export type OptionData<T> = {
  label: string
  value: T
}

export type isMulti = true | false

export interface SelectFieldProps<T> {
  inputId?: string
  className?: string
  isMulti?: isMulti
  isSearchable?: boolean
  isDisabled?: boolean
  menuIsOpen?: boolean
  field?: FieldInputProps<T | T[]>
  'data-testid'?: string
  'aria-invalid'?: boolean
  'aria-required'?: boolean
  closeMenuOnSelect?: boolean
  placeholder?: string
  options: readonly OptionData<T>[]
  value?: T
  form?: FormikProps<unknown>
  name: string
  onSelect?: (
    option: OnChangeValue<OptionData<T>, isMulti>,
    actionMeta: ActionMeta<OptionData<T>>
  ) => void
}

export const SelectField = <T,>({
  className,
  isMulti = false,
  isSearchable = false,
  isDisabled,
  menuIsOpen,
  field,
  form,
  name,
  options,
  value,
  onSelect,
  placeholder,
  closeMenuOnSelect,
  ...props
}: SelectFieldProps<T>): JSX.Element => {
  const onChange = (
    option: SingleValue<OptionData<T>> | MultiValue<OptionData<T>>,
    actionMeta: ActionMeta<OptionData<T>>
  ): void => {
    if (form && field) {
      form.setFieldValue(
        field.name,
        isMulti
          ? (option as MultiValue<
              OptionData<T>
            >)?.map((item: SingleValue<OptionData<T>>) =>
              item ? item.value : null
            )
          : (option as SingleValue<OptionData<T>>)?.value
      )
    }

    if (onSelect === undefined) return

    /* To clear the value with clear indicator detect actionMeta.action === 'clear' in onSelect prop */
    onSelect(option, actionMeta)
  }

  const getValue = (): PropsValue<OptionData<T>> => {
    if (typeof value !== 'undefined') {
      return options.filter(option => option.value === value) ?? null
    }

    if (options && field && form) {
      return isMulti
        ? options.filter(
            option => (field.value as T[]).indexOf(option.value) >= 0
          )
        : options.find(option => option.value === field.value) ??
            (('' as unknown) as SingleValue<OptionData<T>>)
    }

    return null
  }

  return (
    <Select
      menuIsOpen={menuIsOpen}
      isMulti={isMulti}
      isSearchable={isSearchable}
      isDisabled={isDisabled}
      className={classNames('react-select-wrapper', {
        'react-select-error': props?.['aria-invalid']
      })}
      classNamePrefix="react-select-prefix"
      name={name ?? field?.name}
      value={getValue()}
      onChange={onChange}
      options={options}
      placeholder={placeholder}
      closeMenuOnSelect={closeMenuOnSelect}
      components={selectComponents}
      {...props}
    />
  )
}

const DropdownIndicator = <T,>(
  props: DropdownIndicatorProps<OptionData<T>>
) => {
  const { menuIsOpen } = props.selectProps

  return (
    <components.DropdownIndicator {...props}>
      <MenuDown
        className={classNames(
          'w-1.5 h-1 mx-1 text-primary-oxford-blue transform transition-transform',
          {
            'rotate-180': menuIsOpen
          }
        )}
      />
    </components.DropdownIndicator>
  )
}

const selectComponents = { DropdownIndicator, IndicatorSeparator: null }

export default SelectField
