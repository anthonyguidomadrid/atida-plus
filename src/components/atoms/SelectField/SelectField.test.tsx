import React from 'react'
import { render } from '@testing-library/react'
import { openMenu, select } from 'react-select-event'
import SelectField, { OptionData } from './SelectField'
import type { ActionMeta, SingleValue } from 'react-select'
import { Field, Formik } from 'formik'

const OPTIONS: OptionData<string>[] = [
  { label: 'Option 1', value: 'option_1' },
  { label: 'Option 2', value: 'option_2' }
]

const mockedFn = jest.fn()

const renderForm = (select: React.ReactNode) => {
  const result = render(
    <form data-testid="form">
      <label htmlFor="selected_options">Options</label>
      {select}
    </form>
  )
  const form = result.getByTestId('form')
  const input = result.getByLabelText('Options')
  return { ...result, form, input }
}

describe(SelectField, () => {
  it('opens the menu', () => {
    const { getByLabelText, queryByText } = renderForm(
      <SelectField
        options={OPTIONS}
        name="selected_options"
        inputId="selected_options"
        isMulti
      />
    )
    // option is not yet visible
    expect(queryByText('Option 1')).toBeNull()
    openMenu(getByLabelText('Options'))
    // option can now be seen because menu is open
    expect(queryByText('Option 1')).toBeInTheDocument()
  })

  it('selects an option in a single-option input', async () => {
    let value = OPTIONS[0].value
    const { form, rerender } = renderForm(
      <SelectField
        value={value}
        options={OPTIONS}
        name="selected_options"
        inputId="selected_options"
      />
    )

    value = OPTIONS[1].value
    rerender(
      <SelectField
        value={value}
        options={OPTIONS}
        name="selected_options"
        inputId="selected_options"
      />
    )

    expect(form).toMatchSnapshot()
  })

  it('selects several options in a multi-options input', async () => {
    const { form, rerender } = renderForm(
      <SelectField
        value={null}
        isMulti={true}
        options={OPTIONS}
        name="selected_options"
        inputId="selected_options"
      />
    )

    await rerender(
      <SelectField
        value={OPTIONS[0].value}
        isMulti={true}
        options={OPTIONS}
        name="selected_options"
        inputId="selected_options"
      />
    )

    await rerender(
      <SelectField
        value={OPTIONS[1].value}
        isMulti={true}
        options={OPTIONS}
        name="selected_options"
        inputId="selected_options"
      />
    )

    expect(form).toMatchSnapshot()
  })

  it('has selected value', async () => {
    const { form } = renderForm(
      <SelectField
        value={OPTIONS[0].value}
        options={OPTIONS}
        name="selected_options"
        inputId="selected_options"
      />
    )

    expect(form).toHaveFormValues({
      selected_options: 'option_1'
    })
  })

  it('usage with Formik', async () => {
    let optionValue: SingleValue<string> = ''
    let actionMetaValue = null
    const { input, form } = renderForm(
      <Formik<{ selected_options: string }>
        initialValues={{ selected_options: '' }}
        onSubmit={mockedFn}
      >
        <Field
          component={SelectField}
          inputId="selected_options"
          name="selected_options"
          options={OPTIONS}
          onSelect={(
            option: SingleValue<string>,
            actionMeta: ActionMeta<OptionData<string>>
          ) => {
            optionValue = option
            actionMetaValue = actionMeta

            return { optionValue, actionMetaValue }
          }}
        />
      </Formik>
    )

    await select(input, 'Option 1')
    expect(form).toHaveFormValues({ selected_options: 'option_1' })
  })
})
