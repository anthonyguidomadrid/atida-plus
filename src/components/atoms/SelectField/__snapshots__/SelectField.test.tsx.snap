// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField selects an option in a single-option input 1`] = `
<form
  data-testid="form"
>
  <label
    for="selected_options"
  >
    Options
  </label>
  <div
    class="react-select-wrapper css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="react-select-prefix__control css-1s2u09g-control"
    >
      <div
        class="react-select-prefix__value-container react-select-prefix__value-container--has-value css-319lph-ValueContainer"
      >
        <div
          class="react-select-prefix__single-value css-qc6sy-singleValue"
        >
          Option 1
        </div>
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-haspopup="true"
          aria-readonly="true"
          class="css-mohuvp-dummyInput-DummyInput"
          id="selected_options"
          inputmode="none"
          role="combobox"
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="react-select-prefix__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="react-select-prefix__indicator react-select-prefix__dropdown-indicator css-tlfecz-indicatorContainer"
        >
          <svg
            class="w-1.5 h-1 mx-1 text-primary-oxford-blue transform transition-transform"
          />
        </div>
      </div>
    </div>
    <input
      name="selected_options"
      type="hidden"
      value="option_1"
    />
  </div>
</form>
`;

exports[`SelectField selects several options in a multi-options input 1`] = `
<form
  data-testid="form"
>
  <label
    for="selected_options"
  >
    Options
  </label>
  <div
    class="react-select-wrapper css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-5-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="react-select-prefix__control css-1s2u09g-control"
    >
      <div
        class="react-select-prefix__value-container react-select-prefix__value-container--is-multi css-319lph-ValueContainer"
      >
        <div
          class="react-select-prefix__placeholder css-14el2xx-placeholder"
          id="react-select-5-placeholder"
        >
          Select...
        </div>
        <input
          aria-autocomplete="list"
          aria-describedby="react-select-5-placeholder"
          aria-expanded="false"
          aria-haspopup="true"
          aria-readonly="true"
          class="css-mohuvp-dummyInput-DummyInput"
          id="selected_options"
          inputmode="none"
          role="combobox"
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="react-select-prefix__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="react-select-prefix__indicator react-select-prefix__dropdown-indicator css-tlfecz-indicatorContainer"
        >
          <svg
            class="w-1.5 h-1 mx-1 text-primary-oxford-blue transform transition-transform"
          />
        </div>
      </div>
    </div>
    <div>
      <input
        name="selected_options"
        type="hidden"
      />
    </div>
  </div>
</form>
`;
